grunt/tasks/javascript.js

```diff
@@ -131,46 +143,102 @@ module.exports = function(grunt) {
+    const pluginMappings = pluginPackageJSONs.reduce((mappings, packageJSON) => {
+      if (!packageJSON.adapt_framework) return mappings;
+      mappings.map = Object.assign({}, mappings.map, packageJSON.adapt_framework.map);
+      mappings.paths = Object.assign({}, mappings.paths, packageJSON.adapt_framework.paths);
+      return mappings;
+    }, {});
+
     try {
       await restoreCache(cachePath, basePath);
       await cacheManager.clean();
-      const pluginsPath = path.resolve(cwd, options.pluginsPath).replace(convertSlashes, '/');
-
-      // Make src/plugins.js to attach the plugins dynamically
-      if (!fs.existsSync(pluginsPath)) {
-        fs.writeFileSync(pluginsPath, '');
-      }
 
       // Collect all plugin entry points for injection
-      const pluginPaths = [];
-      for (let i = 0, l = options.plugins.length; i < l; i++) {
-        const src = options.plugins[i];
-        grunt.file.expand({
-          filter: options.pluginsFilter
-        }, src).forEach(function(bowerJSONPath) {
-          if (bowerJSONPath === undefined) return;
-          const pluginPath = path.dirname(bowerJSONPath);
-          const bowerJSON = grunt.file.readJSON(bowerJSONPath);
-          const requireJSRootPath = pluginPath.substr(options.baseUrl.length);
-          const requireJSMainPath = path.join(requireJSRootPath, bowerJSON.main);
-          const ext = path.extname(requireJSMainPath);
-          const requireJSMainPathNoExt = requireJSMainPath.slice(0, -ext.length).replace(convertSlashes, '/');
-          pluginPaths.push(requireJSMainPathNoExt);
-        });
-      }
+      const pluginPaths = options.pluginsOrder(pluginsObject.plugins.map(plugin => {
+        if (plugin.name === 'adapt-contrib-core') return null;
+        const requireJSRootPath = plugin.sourcePath.substr(options.baseUrl.length);
+        const requireJSMainPath = path.join(requireJSRootPath, plugin.main);
+        const ext = path.extname(requireJSMainPath);
+        const requireJSMainPathNoExt = requireJSMainPath.slice(0, -ext.length).replace(convertSlashes, '/');
+        return requireJSMainPathNoExt;
+      }).filter(Boolean));
 
       // Collect react templates
       const reactTemplatePaths = [];
       options.reactTemplates.forEach(pattern => {
         grunt.file.expand({
-          filter: options.pluginsFilter
-        }, pattern).forEach(templatePath => reactTemplatePaths.push(templatePath.replace(convertSlashes, '/')));
+          filter: options.pluginsFilter,
+          order: options.pluginsOrder
+        }, pattern).forEach(templatePath => {
+          const requireJSRootPath = templatePath.substr(options.baseUrl.length);
+          return reactTemplatePaths.push(requireJSRootPath.replace(convertSlashes, '/'));
+        });
       });
 
+      const isProduction = false;
+      // These will be overridden/added to in the package.json of any plugin
+      //  they are here for backward compatibility only
+      const v5toV6DefaultMappings = {
```

Comment:
with the maps and paths now specified here can we remove the corresponding config from [scriptLoader](https://github.com/adaptlearning/adapt-contrib-core/blob/master/required/adapt/js/scriptLoader.js)?

Reply:
No. The ones in the scriptLoader are client-side and they happen before adapt.min.js is loaded, these mappings happen at the top of adapt.min.js.

We will be able to remove some of them as we move over to node_modules, specifically: `regenerator-runtime`, `core-js`, `react`, `react-dom`, `html-react-parser` and `semver`. The others will be more difficult.

Reply:
ah, of course, though is the scriptLoader map used at all?

Reply:
The only use I saw was spoor's offline_API_wrapper.js, but I think that has recently been moved into compilation, right?

Reply:
The `map` properties are used heavilly in `adapt-devtools`. The `paths` section is used further down the `scriptLoader.js` to load the foundational libraries+templates and some of those libraries are used by name at various places throughout the core and plugins.

Reply:
Effectively, when core is updated to switch to `node_modules`, we can also drop the `scriptLoader.js` directives, where reasonable to do so considering backward compatibility and load order. `jQuery`, `jqueryMobile`, `underscore`, `backbone`, `handlebars` and `Modernizer` will be most difficult I imagine. `imageReady`, `inview` and `jquery.resize` are libraries I made, so they should be quite easy to integrate into the core. `bowser`, `scrollTo`, `requirejs` and `velocity` I haven't analysed yet.